shared/js/passcode_helper.js

```diff
@@ -0,0 +1,161 @@
+    };
+
+    this._makeDigest = function(pass, salt, iterations, algorithm) {
+      var bytes = this._encode(pass);
+      return crypto.subtle.importKey('raw', bytes, 'PBKDF2', false, [
+        'deriveBits'
+      ]).then((pwKey) => {
+        return this._deriveBits(pwKey, salt, iterations, algorithm);
+      }).catch((error) => {
+        console.error('PasscodeHelper: _derive_bits() failed!', error);
+        throw error;
+      });
+    };
+
+    this.setPasscode = function(newPass) {
+      var lock = navigator.mozSettings.createLock();
+      var getFromSettings = Promise.all([
+        lock.get(SET_DIGEST_ITERATIONS),
+        lock.get(SET_DIGEST_ALGORITHM)
+      ]);
+      var digest = getFromSettings.then((values) => {
+        // Always generate a new salt.
+        _salt = crypto.getRandomValues(new Uint8Array(8));
+        /* the combined lock.get() makes it quite ugly. we get
+         an Array of objects, each with just one key/value,
+         which is the requested setting. let's destruct as follows:
+         */
+        var [digestIteration, digestAlgorithm] = values;
+        _iterations = parseInt(digestIteration[SET_DIGEST_ITERATIONS]) ||
+          DEFAULT_ITERATIONS;
+        _algorithm = typeof(digestAlgorithm[SET_DIGEST_ALGORITHM]) == 'string' ?
+                     values[1][SET_DIGEST_ALGORITHM] : DEFAULT_ALGORITHM;
+
+        return this._makeDigest(newPass, _salt,
+          _iterations, _algorithm)
+          .then(this._storeNewDigest).catch((error) => {
+            console.error('PasscodeHelper: Could not make digest:', error);
+            throw error;
+          });
+      }).catch((error) => {
+        console.error('PasscodeHelper: No Settings?', error);
+        throw error;
+      });
+      return digest;
+    };
+    this._storeNewDigest = function (digest) {
+      // Note: We can now store the salt, since digest was generated!
+      var digestUint8 = new Uint8Array(digest);
+      var newSettings = {};
+      newSettings[SET_DIGEST_SALT] = _salt;
+      newSettings[SET_DIGEST_VALUE] = digestUint8;
+      newSettings[SET_DIGEST_ITERATIONS] = _iterations;
+      newSettings[SET_DIGEST_ALGORITHM] = _algorithm;
+      var lock = navigator.mozSettings.createLock();
+      return lock.set(newSettings).then(() => {
+        return digestUint8;
+      }, (error) => {
+        console.error('PasscodeHelper: Couldnt store new digest');
+        throw error;
+      });
+    };
+    this.checkPasscode = function (testPass) {
+      _testPass = testPass;
+      //get salt & digest out of settings
+      var lock = navigator.mozSettings.createLock();
+      var storedParams = Promise.all([
+        lock.get(SET_DIGEST_SALT),
+        lock.get(SET_DIGEST_ITERATIONS),
+        lock.get(SET_DIGEST_ALGORITHM),
+        lock.get(SET_DIGEST_VALUE)
+      ]);
+      return storedParams.then((values) => {
+        return this._makeAndCompare(values[0], values[1], values[2], values[3]);
```

Comment:
If it's an array and exactly have these elements you can write it as:

    this._makeAndCompare.apply(this, values)

It's even better with ES6 spread operator:

    this._makeAndCompare(...values)